ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -996,12 +1017,74 @@ let rank_hosts_by_best_vgpu ~__context vgpu visible_hosts =
                0L
         )
         hosts
-      |> List.map (fun g -> List.map (fun (h, _) -> h) g)
+      |> List.map (fun g -> List.map fst g)
+
+(* Group all hosts to 2 parts:
+   1. Home server. A list of home server (only one host).
+   2. A list of lists, each list contains hosts with the same count of running VM in the same anti-affinity group.
+      These lists are sorted by VM's count.
+   Combine these lists into one list. The list is like below:
+   [[host1] (home server), [host2, host3] (no VM running), [host4, host4] (one VM running), [host5, host6] (more VMs running), ...]
+*)
+let rank_hosts_by_placement ~__context ~vm ~group =
+  let home_server_hosts =
+    match Db.VM.get_affinity ~__context ~self:vm with
+    | ref when Db.is_valid_ref __context ref ->
+        [ref]
+    | _ ->
+        []
+  in
+  let resident_hosts =
+    Db.VM_group.get_VMs ~__context ~self:group
+    |> List.filter_map (fun vm ->
```

Comment:
How about:
```
|> List.filter_map (fun vm ->
    [
      Db.VM.get_scheduled_to_be_resident_on ~__context ~self:vm
    ; Db.VM.get_resident_on ~__context ~self:vm
    ]
    |> List.find_opt (fun h -> h <> Ref.null)
)
```

Reply:
It may be fine if it's a small group. At the moment, we don't limit the number of VMs in a group. This means all VMs can be placed into one group. And I understand one `Db.VM.*` will cause a DB access which is actually expensive. Here, for each VM, 3 such accesses will be performed.
A usual practice is to get the full record of the VM through one DB access, and read any fields from the record. But the size of the record is big.

Reply:
The nested `if-else` looks not great.